Improve Subscriptions logging

luisramos0 opened the pull request:
#### What? Why?

Progresses #4462 but we will need to check if linking schedules to OCs is fully tracked now.

Add paper_trail to OCs and Schedules and add extra logging to Subscriptions background processes.

Here we do a few different things (I keep it in one PR because this is all related and it makes the review and test process easier, I think):

A - Add paper_trail to OCs and Schedules: this enables us to see what changes were done to these models and when they were created/updated/deleted. We also add a custom_data field to paper_trail versions table so we can have extra info for whatever we need, in this case, we are adding: for Schedules, the OCs they are associated with, and for OCs, the schedules they are associated with. (I thought this is a lighter solution that introducing gem paper_trail-association_tracking).
This addresses the problem in the S2 that led to #4462: when did the user linked the schedule with the OC? I don think it will cover 100% of the cases: I think the user will still be able to associated a schedule to an OC without any trace on the DB. That's why I dont think this issue will close #4462

B - Add different log statements in the proxy orders sync process and in the subs placement and confirmation background jobs so that we know what subs and orders were processed in each run. These statements are not going to solve all our debugging needs but it's a start, these basic ones are specially useful when the process fails to run, we will be able to check the logs and see if it didn't run or if it did run without producing the expected output.

#### What should we test?
This PR can only be tested by a dev and doesnt require a testers to verify anything.

A - We should edit the OC and edit a schedule and verify the table versions contain the previous data on those models.

B - We should setup a subscription or two, edit them and verify that the sync process worked and there are logs mentioning what happened. After that, when the OC opens, we need to verify the placement job worked and there are logs mentioning what happened, including what emails were sent. After that, when the OC closes, we need to verify the confirmation job worked and there are logs mentioning what happened, including what emails were sent.

#### Release notes
Changelog Category: Added
Subscriptions module now has much better logging and it will be easier from now on to debug and explain what happened in specific cases.

sigmundpetersen commented:
Build is failing and takes very long

sigmundpetersen commented:
Build is failing

mkllnk commented on lib/open_food_network/proxy_order_syncer.rb:
```diff
@@ -17,29 +17,36 @@ def initialize(subscriptions)
-    def sync_all!
+    def sync_subscriptions!
       @subscriptions.each do |subscription|
         @subscription = subscription
```
It took me a minute to find this line to understand that your change is correct here. It should really be `sync_subscription!(subscription)` instead of a leaking ivar. Out of scope here though.

luisramos0 replied:
yeah, the use of these ivars is a bit weird in this class...

mkllnk commented on db/schema.rb:
```diff
@@ -1202,6 +1204,7 @@
     t.string   "whodunnit"
     t.text     "object"
     t.datetime "created_at"
+    t.string   "custom_data"
```
This line belongs to your first commit!

luisramos0 replied:
:+1: fixed

luisramos0 commented:
there's something wrong with the timestamps in the order_cycle_schedules... it's breaking the build. I can figure it out...
```

  ActiveRecord::StatementInvalid:
       PG::NotNullViolation: ERROR:  null value in column "created_at" violates not-null constraint
       DETAIL:  Failing row contains (6, 6, 6, null, null).
       : INSERT INTO "order_cycle_schedules" ("schedule_id", "order_cycle_id") VALUES (6, 6) RETURNING "id"
```

any idea?

Matt-Yorkley commented:
I think I saw that whilst playing with the 3-0-stable build as well. The created_at field should set automatically, right? Maybe there's something missing in the schema?

mkllnk commented:
I've never seen this before but here is a theory: We don't have an ActiveRecord class called `OrderCycleSchedule`. The table enables a many-to-many relationship but is not a model on its own. Maybe ActiveRecord doesn't support timestamps on relationship tables. And would it make sense? Each row is either there or not, `updated_at` would never get updated, right?

luisramos0 commented:
For now I have removed the timestamp in OrderCycleSchedules so we can progress with the other changes.
<details>
This is the commit I removed:
![Screenshot 2020-02-07 at 10 06 10](https://user-images.githubusercontent.com/1640378/74020898-6bbfb480-4992-11ea-9226-eff4f63c477a.png)
</details>

This way a user can still link/unlink a schedule from an OC without any trace or log. So the original issue is still open after this PR.

luisramos0 commented:
I need a second approval here before I test, @Matt-Yorkley can you review please?

luisramos0 commented:
ok, this is looking as expected in staging FR.
This is on table versions where new field custom_data holds useful info (although the field is not explicitly identified in the table).

Create OC 35 with a schedule 5:
![image](https://user-images.githubusercontent.com/1640378/74663757-4867e700-5194-11ea-9fc4-318a8671f88c.png)

Add schedule 5 to OC 33:
![image](https://user-images.githubusercontent.com/1640378/74663862-8402b100-5194-11ea-9368-2f7560e10c5e.png)

Create Schedule 5 with OC 34:
![image](https://user-images.githubusercontent.com/1640378/74663912-9977db00-5194-11ea-9fa0-a11945e0db74.png)

Update the name of the schedule will list the OCs linked to it:
![image](https://user-images.githubusercontent.com/1640378/74664046-e065d080-5194-11ea-81df-e208f751d25c.png)

BUT editing a schedule (click on the schedule name) and change the list of OCs, either add or remove OCs, will NOT apppear in teh version DB... this is the only missing use case to close #4462

luisramos0 commented:
Not exactly ideal when the dev tests their own PR.
@Matt-Yorkley or @mkllnk could you please confirm my testing above makes sense and it's clear what remains to be done. If you think all is good, you can merge this PR.

mkllnk commented:
@Matt-Yorkley Can you take this one on?

luisramos0 commented:
well spotted Pau, I had my local db messed up with code from different PRs.
It's fixed in a new commit :+1:

sigmundpetersen commented:
@sauloperez do you approve the last commit? 
Did you want to test this as well afterwards?

sauloperez commented:
All looks good. Thanks @sigmundpetersen . I didn't get to see the logging in all scenarios but I saw some and I've full confidence on what @luisramos0 tested.
